Demo PR for issue #772 get chip temperature from SX127x

cyberman54 opened the pull request:
Seems to work, but not sure if it is safe to call it.

terrillmoore commented:
Hi @cyberman54,

Thanks for the PR.

You're right, it's probably not safe -- if the LMIC is busy doing an uplink, it will probably crash the uplink. The right approach would be to integrate this into the FSM as a request; then the LMIC and the radio driver process when safe and return values. (Another, sneakier approach would be to measure before each uplink, and return that value if the radio is "busy".)

A few other random things.

1. We ought to split this into a radio driver API and an LMIC api. The radio driver's only client is the LMIC; the LMIC (is supposed to) export all the client APIs.
2. We will be supporting the SX126x soon, and therefore the API needs to work in ways that are portable to the 126x (and future radios). So the answer should be in degrees C times some factor (16 or 256 are convenient), and probably should be a s2_t.
3. In case a future radio doesn't support this, we should have a way of returning an error code.
4. It would probably make sense to have another API so the client can set the calibration value. That might be an LMIC-only thing; the radio driver would just focus on getting the value.
5. It would be good to check for non-physical values and not return them (error code instead).
6. I note that their sample code is doing an inefficient form of negation -- I believe that the raw temperature is exactly `-(s1_t)RegTemp`, according to their table and descriptions. (This would have been clear if they had put RegTemp as a signed int8_t in the table in section 5.5.7 of the datasheet. But for some reason they wrote as unsigned, and then came up with the confusing conversion code.)

Best regards,
--Terry

cyberman54 commented:
@terrillmoore Thanks for notes.

> Another, sneakier approach would be to measure before each uplink, and return that value if the radio is "busy".

Where would be a suitable place in radio.c / osmlmic.c / lmic.c to put this in there?

cyberman54 commented:
I now implemented this structure:

radio.c: added `s1_t radio_raw_temp(void)`, which is called by `starttx()`. It reads the radio raw temperature value and stores it in `LMIC.radio.temperature`.

Thus, with every uplink we get an updated temperature reading in `LMIC.radio.temperature`, at zero cost for radio chip duty cycle, because the temperature value is updated only and whenever the transmitter ist started by LMIC.

Client application can use `LMIC.radio.temperature` as API to get the latest value.

/cc @terrillmoore
